src/compiler/scala/tools/nsc/backend/jvm/ClassHandler.scala

```diff
@@ -0,0 +1,298 @@
+    override def endUnit(unit: SourceFile): Unit = ()
+
+    private def globalOptimise(): Unit = {
+      val allClasses = bufferBuilder.result()
+      postProcessor.runGlobalOptimizations(allClasses)
+
+      //replay the units to underlying
+      var current = allClasses.head
+      underlying.startProcess(current)
+      for (next <- allClasses.tail) {
+        if (next.sourceFile ne current.sourceFile) {
+          underlying.endUnit(current.sourceFile)
+        }
+        underlying.startProcess(next)
+        current = next
+      }
+      underlying.endUnit(current.sourceFile)
+    }
+
+    override def initialise(): Unit = {
+      bufferBuilder.clear()
+      underlying.initialise()
+    }
+
+    override def toString: String = s"GloballyOptimising[$underlying]"
+  }
+
+  sealed abstract class WritingClassHandler(val javaExecutor :Executor) extends ClassHandler{
+    val unitInfoLookup: UnitInfoLookup
+    val cfWriter: ClassfileWriter
+
+    protected val pendingBuilder = List.newBuilder[UnitResult]
+    private val inUnit = ListBuffer.empty[GeneratedClass]
+
+    override def startProcess(clazz: GeneratedClass): Unit = {
+      inUnit += clazz
+    }
+
+    protected def pending(): List[UnitResult] = {
+      val result = pendingBuilder.result()
+      pendingBuilder.clear()
+      result
+    }
+    override def initialise(): Unit = {
+      super.initialise()
+      pendingBuilder.clear()
+    }
+
+    override def endUnit(unit:SourceFile): Unit = {
+      val unitProcess = new UnitResult(unitInfoLookup, inUnit.result, unit)
+      inUnit.clear()
+      postProcessUnit(unitProcess)
+
+      pendingBuilder += unitProcess
+    }
+
+    protected implicit val executionContext: ExecutionContextExecutor = ExecutionContext.fromExecutor(javaExecutor)
+
+    final def postProcessUnit(unitProcess: UnitResult): Unit = {
+      unitProcess.task = Future {
+        // we 'take' classes to reduce the memory pressure
+        // as soon as the class is consumed and written, we release its data
+        unitProcess.takeClasses foreach {
+          postProcessor.sendToDisk(unitProcess, _, cfWriter)
+        }
+        unitProcess.completedUnit()
+      }
+    }
+
+    override def close(): Unit = cfWriter.close()
+
+    override def complete(): Unit = {
+      val directBackendReporting = postProcessor.bTypes.frontendAccess.directBackendReporting
+      def stealWhileWaiting(unitResult: UnitResult, fut: Future[Unit]): Unit = {
+        while (!fut.isCompleted)
+          tryStealing match {
+            case Some(r:Runnable) => r.run()
+            case None => Await.ready(fut, Duration.Inf)
+        }
+        //we know that they are complete by we need to check for exception
+        //but first get any reports
+        unitResult.relayReports(directBackendReporting)
+        fut.value.get.get
+      }
+
+      // This way it is easier to test, as the results are deterministic
+      // the the loss of potential performance is probably minimal
```

Comment:
we could process the result in the order of completion, rather than the order of submission

Agreed the comment is poor. Will reword to explain the options